### PR TITLE
StructDecl, UnionDecl and VarDecl

### PR DESCRIPTION
**Pull request by vishnu-m**
[WIP] - For preliminary comments. 
NB : Will merge only after adding tests.

### CODE REVIEW

File: cparser.lua [hunk 1]
```diff
@@ -25,8 +25,58 @@ local function enum_handler(cursor, parent, declarations)
...
+                name = cursor:getSpelling(),
+                type = type:getSpelling(),
+                storage_specifier = cursor:getStorageClass(),  
+        }
+        table.insert(declarations, decl)
+end
+
+local function function_handler(cursor, declarations)
         local func_params = {}
+        local definition = {}
```
Comment (vishnu-m) [inline cparser.lua]: Ignore this @gligneul .

File: cparser.lua [hunk 2]
```diff
@@ -3,9 +3,9 @@ local luaclang = require "luaclang"
 --module table
 local cparser = {}
 
-local function enum_handler(cursor, parent, declarations)
+local function enum_handler(cursor, declarations)
         local fields = {} 
-        cursor:visitChildren(function (cursor, parent)
+        cursor:visitChildren(function (cursor, _)
```
Comment (gligneul) [inline cparser.lua]: In this case you can suppress the parameter given that you don't have a third one.
```suggestion
        cursor:visitChildren(function (cursor)
```

File: cparser.lua [hunk 3]
```diff
@@ -25,8 +25,58 @@ local function enum_handler(cursor, parent, declarations)
...
 end
 
-local function function_handler(cursor, parent, declarations)
+local function union_struct_handler(cursor, declarations) 
+        local fields = {}
+        local definition = {}
+        local decl = {
+                name = cursor:getSpelling()
+        }
+        if cursor:getKind() == "StructDecl" then
```
Comment (gligneul) [inline cparser.lua]: This is a nice place to use Lua ternary operator
```suggestion
        local tag = cursor:getKind() == "StructDecl" and 'struct' or 'union'
```

File: cparser.lua [hunk 4]
```diff
@@ -25,8 +25,58 @@ local function enum_handler(cursor, parent, declarations)
         table.insert(declarations, decl)
 end
 
-local function function_handler(cursor, parent, declarations)
+local function union_struct_handler(cursor, declarations) 
+        local fields = {}
+        local definition = {}
+        local decl = {
```
Comment (gligneul) [inline cparser.lua]: I think you should create the declaration in the end just like the other handlers.

File: cparser.lua [hunk 5]
```diff
@@ -25,8 +25,58 @@ local function enum_handler(cursor, parent, declarations)
         table.insert(declarations, decl)
 end
 
-local function function_handler(cursor, parent, declarations)
+local function union_struct_handler(cursor, declarations) 
+        local fields = {}
+        local definition = {}
```
Comment (gligneul) [inline cparser.lua]: I would rename it to be consistent with the current vocabulary that we are using. e.g. Clang calls struct definitions StructDecl.
```suggestion
        local chield_declarations = {}
```
Another option would be inserting all declarations in the top-level. This option might be better even though we lose a bit of info.

File: cparser.lua [hunk 6]
```diff
@@ -25,8 +25,58 @@ local function enum_handler(cursor, parent, declarations)
...
+        local decl = {
+                name = cursor:getSpelling()
+        }
+        if cursor:getKind() == "StructDecl" then
+                decl.tag = 'struct'
+        else 
+                decl.tag = 'union'
+        end
+        cursor:visitChildren(function (cursor, _)
```
Comment (gligneul) [inline cparser.lua]:
```suggestion
        cursor:visitChildren(function (cursor)
```

**Review by gligneul — CHANGES_REQUESTED** [review]

**Review by gligneul — APPROVED** [review]

**Pull request merged**